### PR TITLE
Reject abstract types in `var` function parameters

### LINKED ISSUE #6402: Missing error for abstract type in `var` parameter of function definition
### Description of the bug:

The toolchain accepts `var` parameters with abstract types in function definitions.

### What did you do, or what's a simple way to reproduce the bug?

```carbon
abstract class A {}
fn F(var a: A) {}
```

### What did you expect to happen?

This example should produce an error, as `var`s with abstract types do in other contexts.

### What actually happened?

Example is accepted; code is generated for `F`. However, it doesn't seem to be possible to call `F`, as any call would create an instance of an `A`.

### Any other information, logs, or outputs that you want to share?

We should also consider whether we want to reject abstract classes as parameter types in forward declarations. It seems reasonable to reject:

```carbon
abstract class A;
fn F(var a: A);
```

but for a class imported from C++, we don't necessarily know whether it is abstract when we see a forward declaration:

```carbon
import Cpp inline '''
struct A;
struct B; // #1
struct B { virtual void f() = 0; }; // #2
''';

// Presumably should be OK, we don't know whether A is abstract.
fn F(var a: Cpp.A);
// Should we reject this because B happens to be complete? Or accept,
// as we would if we had only seen #1 and not #2?
fn G(var a: Cpp.B);
```

### PR DESCRIPTION
**Pull request by ozguronsoy**
### Description
Fixes an issue where the toolchain accepted abstract types in function parameters declared with `var`.

### Changes
- Implemented a check for abstract types for function parameters with `var` binding pattern in `HandleAnyBindingPattern`.
- Added a test case to `toolchain/check/testdata/class/fail_abstract.carbon`.

**Note:** I did not use `AsConcreteType` like used in `case FullPatternStack::Kind::NameBindingDecl`. Using it enforces type completion, thus causing valid signatures such as `fn F[var self: Self]` to fail.

Also the pre-commit checks fail due to a diagnostic name collision with `toolchain/check/type_completion.cpp`. Should I add a function that just checks if the type is abstract to share the diagnostic?

Fixes #6402

### CODE REVIEW

File: toolchain/check/handle_binding_pattern.cpp [hunk 1]
```diff
@@ -192,6 +201,29 @@ static auto HandleAnyBindingPattern(Context& context, Parse::NodeId node_id,
...
+            context.types().GetUnqualifiedType(cast_type_id);
+        if (context.types().Is<SemIR::ClassType>(unqualified_type_id)) {
+          auto class_type =
+              context.types().GetAs<SemIR::ClassType>(unqualified_type_id);
+          auto& class_info = context.classes().Get(class_type.class_id);
+          if (class_info.inheritance_kind ==
+              SemIR::Class::InheritanceKind::Abstract) {
+            auto builder = abstract_diagnoser();
+            CARBON_DIAGNOSTIC(
+                ClassAbstractHere, Note,
```
Comment (jonmeow) [inline toolchain/check/handle_binding_pattern.cpp]: We don't copy-paste diagnostics, and you'll have a unit test failure because of the `ClassAbstractHere` duplication. Had you considered instead exposing `NoteAbstractClass` and calling it?

**Review by jonmeow — COMMENTED** [review]
Thanks! This looks helpful, and should just need a few fixes.

I'm running workflows expecting them to show failures for you to look at, but you can equivalents locally:

- For tests, `bazel test //toolchain/...` will include the diagnostic test we use to avoid duplicate diagnostics
- For pre-commit, [see here](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/project/contribution_tools.md#running-pre-commit).

File: toolchain/check/handle_binding_pattern.cpp [hunk 2]
```diff
@@ -192,6 +201,29 @@ static auto HandleAnyBindingPattern(Context& context, Parse::NodeId node_id,
         break;
       }
 
+      // Using `AsConcreteType` here causes `fn F[var self: Self]();`
+      // to fail since `Self` is an incomplete type.
+      if (node_kind == Parse::NodeKind::VarBindingPattern) {
+        auto unqualified_type_id =
+            context.types().GetUnqualifiedType(cast_type_id);
```
Comment (jonmeow) [inline toolchain/check/handle_binding_pattern.cpp]: Actually, a `partial C` for `abstract C` should be valid here -- you might need to include qualifiers

(partial types are always complete (if defined))

File: toolchain/check/handle_binding_pattern.cpp [hunk 3]
```diff
@@ -22,15 +22,15 @@
 
 namespace Carbon::Check {
 
-auto HandleParseNode(Context& context, Parse::UnderscoreNameId node_id)
-    -> bool {
+auto HandleParseNode(Context& context,
+                     Parse::UnderscoreNameId node_id) -> bool {
```
Comment (CarbonInfraBot) [inline toolchain/check/handle_binding_pattern.cpp]: **[diff]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>
```suggestion
auto HandleParseNode(Context& context, Parse::UnderscoreNameId node_id)
    -> bool {
```

File: toolchain/check/handle_binding_pattern.cpp [hunk 4]
```diff
@@ -22,15 +22,15 @@
...
   context.node_stack().Push(node_id, SemIR::NameId::Underscore);
   return true;
 }
 
 // Returns the `InstKind` corresponding to the pattern's `NodeKind`.
-static auto GetPatternInstKind(Parse::NodeKind node_kind, bool is_ref)
-    -> SemIR::InstKind {
+static auto GetPatternInstKind(Parse::NodeKind node_kind,
+                               bool is_ref) -> SemIR::InstKind {
```
Comment (CarbonInfraBot) [inline toolchain/check/handle_binding_pattern.cpp]: **[diff]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>
```suggestion
static auto GetPatternInstKind(Parse::NodeKind node_kind, bool is_ref)
    -> SemIR::InstKind {
```

File: toolchain/check/handle_binding_pattern.cpp [hunk 5]
```diff
@@ -265,14 +289,14 @@ static auto HandleAnyBindingPattern(Context& context, Parse::NodeId node_id,
   return true;
 }
 
-auto HandleParseNode(Context& context, Parse::LetBindingPatternId node_id)
-    -> bool {
+auto HandleParseNode(Context& context,
+                     Parse::LetBindingPatternId node_id) -> bool {
```
Comment (CarbonInfraBot) [inline toolchain/check/handle_binding_pattern.cpp]: **[diff]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>
```suggestion
auto HandleParseNode(Context& context, Parse::LetBindingPatternId node_id)
    -> bool {
```

File: toolchain/check/handle_binding_pattern.cpp [hunk 6]
```diff
@@ -265,14 +289,14 @@ static auto HandleAnyBindingPattern(Context& context, Parse::NodeId node_id,
...
   return HandleAnyBindingPattern(context, node_id,
                                  Parse::NodeKind::LetBindingPattern);
 }
 
-auto HandleParseNode(Context& context, Parse::VarBindingPatternId node_id)
-    -> bool {
+auto HandleParseNode(Context& context,
+                     Parse::VarBindingPatternId node_id) -> bool {
```
Comment (CarbonInfraBot) [inline toolchain/check/handle_binding_pattern.cpp]: **[diff]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>
```suggestion
auto HandleParseNode(Context& context, Parse::VarBindingPatternId node_id)
    -> bool {
```

File: toolchain/check/handle_binding_pattern.cpp [hunk 7]
```diff
@@ -372,8 +396,8 @@ auto HandleParseNode(Context& context,
   return true;
 }
 
-auto HandleParseNode(Context& context, Parse::FieldNameAndTypeId node_id)
-    -> bool {
+auto HandleParseNode(Context& context,
+                     Parse::FieldNameAndTypeId node_id) -> bool {
```
Comment (CarbonInfraBot) [inline toolchain/check/handle_binding_pattern.cpp]: **[diff]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>
```suggestion
auto HandleParseNode(Context& context, Parse::FieldNameAndTypeId node_id)
    -> bool {
```

File: toolchain/check/handle_binding_pattern.cpp [hunk 8]
```diff
@@ -420,14 +444,14 @@ auto HandleParseNode(Context& context, Parse::FieldNameAndTypeId node_id)
   return true;
 }
 
-auto HandleParseNode(Context& context, Parse::RefBindingNameId node_id)
-    -> bool {
+auto HandleParseNode(Context& context,
+                     Parse::RefBindingNameId node_id) -> bool {
```
Comment (CarbonInfraBot) [inline toolchain/check/handle_binding_pattern.cpp]: **[diff]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>
```suggestion
auto HandleParseNode(Context& context, Parse::RefBindingNameId node_id)
    -> bool {
```

File: toolchain/check/handle_binding_pattern.cpp [hunk 9]
```diff
@@ -420,14 +444,14 @@ auto HandleParseNode(Context& context, Parse::FieldNameAndTypeId node_id)
...
   context.node_stack().Push(node_id);
   return true;
 }
 
-auto HandleParseNode(Context& context, Parse::TemplateBindingNameId node_id)
-    -> bool {
+auto HandleParseNode(Context& context,
+                     Parse::TemplateBindingNameId node_id) -> bool {
```
Comment (CarbonInfraBot) [inline toolchain/check/handle_binding_pattern.cpp]: **[diff]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>
```suggestion
auto HandleParseNode(Context& context, Parse::TemplateBindingNameId node_id)
    -> bool {
```

**Comment by ozguronsoy** [comment]
Thanks, the issues should be fixed now. I also added a test for `fn F(var p: partial Abstract)` signature.

**Review by jonmeow — APPROVED** [review]
Looks good, thanks for the contribution!
